modules/prometheus/src/main/scala/com/evolutiongaming/smetrics/Prometheus.scala

```diff
@@ -0,0 +1,244 @@
+package com.evolutiongaming.smetrics
+
+import java.io.StringWriter
+
+import cats.effect.{Resource, Sync}
+import cats.implicits._
+import io.prometheus.client.Collector
+import io.prometheus.client.exporter.common.TextFormat
+import io.prometheus.{client => P}
+
+trait Prometheus[F[_]] {
+
+  val registry: CollectorRegistry[F]
+
+  val write004: F[String]
+}
+
+object Prometheus {
+
+  def apply[F[_] : Sync](
+    collectorRegistry: P.CollectorRegistry
+  ): Prometheus[F] = {
+
+    def register[A <: Collector](collector: A): Resource[F, A] = {
+      val result = for {
+        _ <- Sync[F].delay {
+          collectorRegistry.register(collector)
+        }
+      } yield {
+        val release = Sync[F].delay {
+          collectorRegistry.unregister(collector)
+        }
+        (collector, release)
+      }
+      Resource(result)
+    }
+
+    def apply[A, B[_], Child, C <: P.SimpleCollector[Child], Builder <: P.SimpleCollector.Builder[Builder, C], R](
+      builder: Builder, labels: A
+    )(
+      implicit
+      magnet: LabelsMagnet[A, B],
+      fromCollector: C => R,
+      fromCollectorChild: Child => R
+    ): Resource[F, B[R]] = {
+
+      val labelNames = magnet.names(labels)
+      val collector = builder.labelNames(labelNames: _*).create()
+      for {
+        collector <- register(collector)
+      } yield {
+        magnet.withValues { labelValues =>
+          if (labelValues.isEmpty) collector.as[R]
+          else collector.labels(labelValues: _*).as[R]
+        }
+      }
+    }
+
+    new Prometheus[F] {
+
+      val registry: CollectorRegistry[F] = new CollectorRegistry[F] {
```

Comment:
I think that we can leave CollectorRegistryPrometheus as is and just use it here, so:

the diff is smaller
history is still in place
this way it is the most flexible
you won't even break backward compatibli